### PR TITLE
Initial update to CDKv2

### PR DESCRIPTION
**Pull request by jgreenbow-userful**
- updated requirements files
- replacement of aws_cdk.core with explicit imports
- use of constructs module
- removal of pylint useless option values

### CODE REVIEW

**Comment by jgreenbow-userful** [comment]
Issue: does not pass tests due to incompatibilities between cdk-chalice (CDKv1 only) and chalice[cdkv2] implementation. Someone with experience in chalice could take it from here, or we could drop the test for now.

**Comment by jgreenbow-userful** [comment]
EDIT: snipped error log to file

`$ ./scripts/run-tests.sh`

[aws-cdk-project-structure-python_run-tests_01.log](https://github.com/aws-samples/aws-cdk-project-structure-python/files/8928790/aws-cdk-project-structure-python_run-tests_01.log)

**Comment by alexpulver** [comment]
Hi @jgreenbow-userful, thanks for the work and submitting the pull request! Since this code supports an already published blog post, I sadly can't merge these changes, as it will require changes to the blog post itself. Mentioning just in case - cdk-chalice and chalice[cdk]/chalice[cdkv2] are two different implementations, and replace each other. For the blog post this code supports, it doesn't matter much which one to use, as it doesn't affect the recommendation on project structure. When I bumped pylint version in another project of mine, I also suddenly saw the first two options becoming a no-op, although they worked before. Please let me know if you have any thoughts on that, and again, thank you for submitting the pull request!

**Comment by alexpulver** [comment]
P.S.: I am looking into supporting CDK v2 in cdk-chalice: https://github.com/alexpulver/cdk-chalice/issues/93

**Comment by jgreenbow-userful** [comment]
@alexpulver That all makes sense, and I should add that I appreciate your original recommendations and project structure.

That being said, this repo is quite a bit less useful if it needs significant modification to update before it can be used. How can users get access to a CDKv2 version? A new repo in aws-samples? A branch in this repo and remark in the blog post or README? 

Just wondering what options exist to make this project useful to future CDKers. Thank you again, it has been a great launch point!

**Comment by jgreenbow-userful** [comment]
I also suppose the answer may have to wait until cdk2-chalice gets merged.

Would you consider merging this pull request into a branch other than main? Or shall we close it?

**Comment by alexpulver** [comment]
Thank you for the kind words 🤗. I liked your idea of a separate branch - created `chalice-cdkv2` branch and made it to be the base branch for the pull request. Once I merge the pull request, I'll add a note about the new branch in the `README.md` file on the `main` branch. Can you please add the fact of migrating from cdk-chalice library to chalice[cdkv2] in the commit description?

**Comment by jgreenbow-userful** [comment]
👍 Updated the commit description.

File: app.py
Comment (alexpulver) [inline app.py]: I'd prefer to be explicit about `App`'s module, since this is quite a generic word:
```suggestion
import aws_cdk as cdk
```

File: constants.py
Comment (alexpulver) [inline constants.py]: I'd prefer to be explicit about `Environment`'s module, since this is quite a generic word:
```suggestion
import aws_cdk as cdk
```

File: database/infrastructure.py
Comment (alexpulver) [inline database/infrastructure.py]: I'd prefer to keep module namespace. That should allow a reader to quicker understand which module `RemovalPolicy` belongs to:
```suggestion
import aws_cdk as cdk
```

File: deployment.py
Comment (alexpulver) [inline deployment.py]: I'd prefer to keep the module namespace for a reader to quicker understand which module `Stack` and `Stage` classes come from, as there is no "natural" mapping in my mind. That is unlike `Construct` which is very generic:
```suggestion
import aws_cdk as cdk
```

File: deployment.py
Comment (alexpulver) [inline deployment.py:22]: The reason I did a direct import here is that these are class-containing modules, per https://peps.python.org/pep-0008/#imports

File: monitoring/infrastructure.py
Comment (alexpulver) [inline monitoring/infrastructure.py]: I'd prefer to keep the module namespace for the reader to quicker understand which module they belong to:
```suggestion
import aws_cdk as cdk
```

File: monitoring/infrastructure.py
Comment (alexpulver) [inline monitoring/infrastructure.py:33]: I usually avoid adding type information in variable name. Is there any specific reason you wanted to change this? If you think it is needed, perhaps a type hint will do?

```suggestion
            dimensions: Dict[str,str] =apigateway_metric_dimensions,
```

File: api/runtime/requirements.txt
Comment (alexpulver) [inline api/runtime/requirements.txt]: I see you used Python 3.10, but this project uses 3.7 (see the instructions in README.md) to extend reach for customers that didn't yet upgrade. Can you please use Python 3.7 for upgrading dependencies and validating code changes?

File: requirements-dev.txt
Comment (alexpulver) [inline requirements-dev.txt]: Same comment here as for `api/runtime/requirements.txt` - can you please use Python 3.7?

File: requirements.in
Comment (alexpulver) [inline requirements.in]: This library doesn't have a version lower than 2.x, so this is a no-op and can confuse people to think why there is a version specifier here:
```suggestion
aws-cdk-lib
```

File: requirements.in
Comment (alexpulver) [inline requirements.in]: aws-cdk-lib has `"constructs": "^10.0.0"` in its [peerDependencies](https://github.com/aws/aws-cdk/blob/v2-release/packages/aws-cdk-lib/package.json#L369), so version specifier should not be required.
```suggestion
constructs
```

File: requirements.txt
Comment (alexpulver) [inline requirements.txt]: Same comment here as for `api/runtime/requirements.txt` - can you please use Python 3.7?

File: tests/test_api_infrastructure.py
Comment (alexpulver) [inline tests/test_api_infrastructure.py]: I'd prefer to be explicit about `App`'s and `Stack`'s module, based on the same reasoning as in prior comments:
```suggestion
import aws_cdk as cdk
```

File: pipeline.py
Comment (alexpulver) [inline pipeline.py]: I'd prefer to keep module namespace for the reader to quicker understand which module `Stack` belongs to:
```suggestion
import aws_cdk as cdk
```

**Review by alexpulver — CHANGES_REQUESTED** [review]
Thanks! I have a few comments which are mostly cosmetic - can you please take a look?

File: monitoring/infrastructure.py
Comment (jgreenbow-userful) [reply monitoring/infrastructure.py:33]: In this case, `dimensions` was deprecated from v1 -> v2. `dimensions_map` is the only option. https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_cloudwatch/Metric.html

File: api/runtime/requirements.txt
Comment (jgreenbow-userful) [reply api/runtime/requirements.txt]: Installed Python 3.7 from ppa:deadsnakes

File: requirements.in
Comment (jgreenbow-userful) [reply requirements.in]: Removed no-op

File: requirements.in
Comment (jgreenbow-userful) [reply requirements.in]: Removed no-op

**Comment by alexpulver** [comment]
Hi @jgreenbow-userful, thanks for submitting the changes! I am on vacation until September, hence didn't look into the review yet. I will go over that when I return.

**Comment by jgreenbow-userful** [comment]
No worries @alexpulver , I have been enjoying some vacation time myself. This submission is not in any critical path, so no reason for urgency. Make the most of your time away!

File: requirements-dev.in
Comment (alexpulver) [inline requirements-dev.in:5]: Can you please add a comment about the reason for pinning to 3.9.x? (is it because of a conflict with `importlib-metadata` versions otherwise?)

**Review by alexpulver — CHANGES_REQUESTED** [review]
One last comment and we are done 😃

File: requirements-dev.in
Comment (jgreenbow-userful) [reply requirements-dev.in:5]: Sure, that change was in commit [0b7d1f6](https://github.com/jgreenbow-userful/aws-cdk-project-structure-python/commit/0b7d1f6c82d42698245806418c449ba8c6081bca). From that commit message:

> Pinned flake8 to prevent dependency conflict
>
> flake8>=4.0 pins importlib-metadata<4.3 as a dependency. This is incompatible
> with other packages requiring importlib-metadata>=4.12.

This is a Python 3.7 issue, as I believe flake8 for Python>=3.8 does not have the same dependency lock. 

However, I will do a quick re-run on for pip-compile to see if this has been patched.

Comment (jgreenbow-userful) [reply requirements-dev.in:5]: I just ran another `pip-compile` on my Python 3.7.14 but same issue. Here is the output.

`> pip-compile --upgrade requirements-dev.in`
<details>
Could not find a version that matches importlib-metadata<4.3,==4.12.0,>=1.1.0 (from -c requirements.txt (line 41))
Tried: 0.0, 0.1, 0.1, 0.2, 0.2, 0.3, 0.3, 0.4, 0.4, 0.5, 0.5, 0.6, 0.6, 0.7, 0.7, 0.8, 0.8, 0.9, 0.9, 0.10, 0.10, 0.11, 0.11, 0.12, 0.12, 0.13, 0.13, 0.14, 0.14, 0.15, 0.15, 0.16, 0.16, 0.17, 0.17, 0.18, 0.18, 0.19, 0.19, 0.20, 0.20, 0.21, 0.21, 0.22, 0.22, 0.23, 0.23, 1.0.0, 1.0.0, 1.1.0, 1.1.0, 1.1.1, 1.1.1, 1.1.2, 1.1.2, 1.1.3, 1.1.3, 1.2.0, 1.2.0, 1.3.0, 1.3.0, 1.4.0, 1.4.0, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 2.0.0, 2.0.0, 2.1.0, 2.1.0, 2.1.1, 2.1.1, 2.1.2, 2.1.2, 2.1.3, 2.1.3, 3.0.0, 3.0.0, 3.1.0, 3.1.0, 3.1.1, 3.1.1, 3.2.0, 3.2.0, 3.3.0, 3.3.0, 3.4.0, 3.4.0, 3.5.0, 3.5.0, 3.6.0, 3.6.0, 3.7.0, 3.7.0, 3.7.1, 3.7.1, 3.7.2, 3.7.2, 3.7.3, 3.7.3, 3.8.0, 3.8.0, 3.8.1, 3.8.1, 3.8.2, 3.8.2, 3.9.0, 3.9.0, 3.9.1, 3.9.1, 3.10.0, 3.10.0, 3.10.1, 3.10.1, 4.0.0, 4.0.0, 4.0.1, 4.0.1, 4.1.0, 4.1.0, 4.2.0, 4.2.0, 4.3.0, 4.3.0, 4.3.1, 4.3.1, 4.4.0, 4.4.0, 4.5.0, 4.5.0, 4.6.0, 4.6.0, 4.6.1, 4.6.1, 4.6.2, 4.6.2, 4.6.3, 4.6.3, 4.6.4, 4.6.4, 4.7.0, 4.7.0, 4.7.1, 4.7.1, 4.8.0, 4.8.0, 4.8.1, 4.8.1, 4.8.2, 4.8.2, 4.8.3, 4.8.3, 4.9.0, 4.9.0, 4.10.0, 4.10.0, 4.10.1, 4.10.1, 4.11.0, 4.11.0, 4.11.1, 4.11.1, 4.11.2, 4.11.2, 4.11.3, 4.11.3, 4.11.4, 4.11.4, 4.12.0, 4.12.0  

There are incompatible versions in the resolved dependencies:  
  importlib-metadata==4.12.0 (from -c requirements.txt (line 41))  
  importlib-metadata<4.3,>=1.1.0 (from flake8==5.0.4->-r requirements-dev.in (line 5))  
</details>

Would you prefer to pin `flake8` at a lower version?

Comment (alexpulver) [reply requirements-dev.in:5]: Thanks for checking! I thought that this is the issue, as I had it as well when bumping dependencies. I feel we should still support Python 3.7, for example, because [Lambda runtime](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html) still supports it. So let's keep the way you pinned flake8 in `requirements-dev.in`. Git blame will help to explain as you mentioned 😃.

image

**Review by alexpulver — APPROVED** [review]

**Comment by alexpulver** [comment]
@jgreenbow-userful thanks a lot for your work on this! We further discussed internally, and want to update the blog post and the trunk branch implementation to AWS CDK v2 using your commit. I anyway wanted to update the blog post and this project with some changes to recommended project structure and internal implementation, so it is a great opportunity to do both together.

**Comment by jgreenbow-userful** [comment]
@alexpulver Thanks so much, that's great to hear!

**Comment by alexpulver** [comment]
@jgreenbow-userful FYI I updated the mainline and the blog post with AWS CDK v2 + some additional changes to project structure and the component we wanted to do. Thanks again for your contribution!

**Pull request merged**
